// @aws-amplify/api-graphql/dist/cjs/internals/generateClient.js
'use strict';

Object.defineProperty(exports, "__esModule", { value: true });
exports.generateClient = void 0;
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
const v6_1 = require("./v6");
const generateModelsProperty_1 = require("./generateModelsProperty");
const types_1 = require("../types");
const core_1 = require("@aws-amplify/core");
/**
 * @private
 *
 * Creates a client that can be used to make GraphQL requests, using a provided `AmplifyClassV6`
 * compatible context object for config and auth fetching.
 *
 * @param params
 * @returns
 */
function generateClient(params) {
    const client = {
        [types_1.__amplify]: params.amplify,
        [types_1.__authMode]: params.authMode,
        [types_1.__authToken]: params.authToken,
        [types_1.__headers]: params.headers,
        graphql: v6_1.graphql,
        cancel: v6_1.cancel,
        isCancelError: v6_1.isCancelError,
        models: {},
    };
    const config = params.amplify.getConfig();
    if (!config.API?.GraphQL) {
        // This happens when the `Amplify.configure()` call gets evaluated after the `generateClient()` call.
        //
        // Cause: when the `generateClient()` and the `Amplify.configure()` calls are located in
        // different source files, script bundlers may randomly arrange their orders in the production
        // bundle.
        //
        // With the current implementation, the `client.models` instance created by `generateClient()`
        // will be rebuilt on every `Amplify.configure()` call that's provided with a valid GraphQL
        // provider configuration.
        //
        // TODO: revisit, and reverify this approach when enabling multiple clients for multi-endpoints
        // configuration.
        client.models = emptyModels;
        generateModelsPropertyOnAmplifyConfigure(client);
    }
    else {
        client.models = (0, generateModelsProperty_1.generateModelsProperty)(client, config.API?.GraphQL);
    }
    return client;
}
exports.generateClient = generateClient;
const generateModelsPropertyOnAmplifyConfigure = (clientRef) => {
    core_1.Hub.listen('core', coreEvent => {
        if (!isConfigureEvent(coreEvent.payload)) {
            return;
        }
        const { data: resourceConfig } = coreEvent.payload;
        if (resourceConfig.API?.GraphQL) {
            clientRef.models = (0, generateModelsProperty_1.generateModelsProperty)(clientRef, resourceConfig.API?.GraphQL);
        }
    });
};
function isConfigureEvent(payload) {
    return payload.event === 'configure';
}
const emptyModels = new Proxy({}, {
    get() {
        throw new Error('Could not generate client. This is likely due to Amplify.configure() not being called prior to generateClient().');
    },
});
